File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/ui/streams/UserMovieStreamFragment.java
package com.battlelancer.seriesguide.ui.streams;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.view.View;
import android.widget.ListAdapter;
import com.battlelancer.seriesguide.adapters.MovieHistoryAdapter;
import com.battlelancer.seriesguide.loaders.TraktMovieHistoryLoader;
import com.battlelancer.seriesguide.ui.HistoryActivity;
import com.battlelancer.seriesguide.ui.MovieDetailsActivity;
import com.uwetrottmann.trakt5.entities.HistoryEntry;

/**
 * Displays a stream of movies the user has recently watched on trakt.
 */
public class UserMovieStreamFragment extends StreamFragment {

    private MovieHistoryAdapter adapter;

    @Override
    protected ListAdapter getListAdapter() {
        if (adapter == null) {
            adapter = new MovieHistoryAdapter(getActivity(), itemClickListener);
        }
        return adapter;
    }

    @Override
    protected void initializeStream() {
        getLoaderManager().initLoader(HistoryActivity.MOVIES_LOADER_ID, null,
                activityLoaderCallbacks);
    }

    @Override
    protected void refreshStream() {
        getLoaderManager().restartLoader(HistoryActivity.MOVIES_LOADER_ID, null,
                activityLoaderCallbacks);
    }

    private MovieHistoryAdapter.OnItemClickListener itemClickListener
            = new MovieHistoryAdapter.OnItemClickListener() {
        @Override
        public void onItemClick(View view, HistoryEntry item) {
            if (item == null) {
                return;
            }

            // display movie details
            if (item.movie == null || item.movie.ids == null) {
                return;
            }
            Intent i = MovieDetailsActivity.intentMovie(getActivity(), item.movie.ids.tmdb);

            ActivityCompat.startActivity(getActivity(), i,
                    ActivityOptionsCompat
                            .makeScaleUpAnimation(view, 0, 0, view.getWidth(), view.getHeight())
                            .toBundle()
            );
        }
    };

    private LoaderManager.LoaderCallbacks<TraktMovieHistoryLoader.Result> activityLoaderCallbacks =
            new LoaderManager.LoaderCallbacks<TraktMovieHistoryLoader.Result>() {
                @Override
                public Loader<TraktMovieHistoryLoader.Result> onCreateLoader(int id, Bundle args) {
                    showProgressBar(true);
                    return new TraktMovieHistoryLoader(getActivity());
                }

                @Override
                public void onLoadFinished(Loader<TraktMovieHistoryLoader.Result> loader,
                        TraktMovieHistoryLoader.Result data) {
                    if (!isAdded()) {
                        return;
                    }
                    adapter.setData(data.results);
                    setEmptyMessage(data.emptyText);
                    showProgressBar(false);
                }

                @Override
                public void onLoaderReset(Loader<TraktMovieHistoryLoader.Result> loader) {
                    // keep current data
                }
            };
}

File: app/src/main/java/com/battlelancer/seriesguide/history/UserMovieStreamFragment.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2014-2024 Uwe Trottmann

package com.battlelancer.seriesguide.history

import android.os.Bundle
import android.view.View
import androidx.loader.app.LoaderManager
import androidx.loader.content.Loader
import com.battlelancer.seriesguide.movies.details.MovieDetailsActivity
import com.battlelancer.seriesguide.util.startActivityWithAnimation
import com.uwetrottmann.trakt5.entities.HistoryEntry

/**
 * Displays a stream of movies the user has recently watched on Trakt.
 */
class UserMovieStreamFragment : StreamFragment() {

    private var adapter: MovieHistoryAdapter? = null

    override val listAdapter: BaseHistoryAdapter
        get() {
            if (adapter == null) {
                adapter = MovieHistoryAdapter(requireContext(), itemClickListener)
            }
            return adapter!!
        }

    override fun initializeStream() {
        LoaderManager.getInstance(this).initLoader(
            HistoryActivity.MOVIES_LOADER_ID, null,
            activityLoaderCallbacks
        )
    }

    override fun refreshStream() {
        LoaderManager.getInstance(this).restartLoader(
            HistoryActivity.MOVIES_LOADER_ID, null,
            activityLoaderCallbacks
        )
    }

    private val itemClickListener = object : BaseHistoryAdapter.OnItemClickListener {
        override fun onItemClick(view: View, item: HistoryEntry) {
            // display movie details
            val tmdb = item.movie?.ids?.tmdb ?: return
            val i = MovieDetailsActivity.intentMovie(requireContext(), tmdb)
            requireContext().startActivityWithAnimation(i, view)
        }
    }

    private val activityLoaderCallbacks: LoaderManager.LoaderCallbacks<TraktEpisodeHistoryLoader.Result> =
        object : LoaderManager.LoaderCallbacks<TraktEpisodeHistoryLoader.Result> {
            override fun onCreateLoader(
                id: Int,
                args: Bundle?
            ): Loader<TraktEpisodeHistoryLoader.Result> {
                showProgressBar(true)
                return TraktMovieHistoryLoader(requireContext())
            }

            override fun onLoadFinished(
                loader: Loader<TraktEpisodeHistoryLoader.Result>,
                data: TraktEpisodeHistoryLoader.Result
            ) {
                setListData(data.results, data.emptyText)
            }

            override fun onLoaderReset(loader: Loader<TraktEpisodeHistoryLoader.Result>) {
                // keep current data
            }
        }
}